### PR TITLE
Add FinishLex function for explicitly placing EOF, abstract artifact superclasses

### PR DESCRIPTION
**Pull request by jackoalan**
This modifies the lexer interface for more flexible source input with significantly reduced string buffering.

Also made `IntermediateLanguageWriter` part of the MonC library as `ILListingWriter` which can be conveniently invoked via the abstract `ModuleArtifact` API.

### CODE REVIEW

File: src/MonC/Artifact.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+﻿using System;
+using System.IO;
+
+namespace MonC
+{
+    /// <summary>
+    /// Abstract compiler artifact which may potentially own unmanaged resources
+    /// </summary>
+    public abstract class Artifact : IDisposable
```
Comment (sfuller) [inline src/MonC/Artifact.cs]: This might be better suited as an interface, as there's no base implementation here.

File: src/MonC/Artifact.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        public virtual void WriteListing(TextWriter writer)
+        {
+        }
+    }
+
+    /// <summary>
+    /// Abstract compiler artifact intended to be used as VM input
+    /// </summary>
+    public abstract class ExecutableArtifact : Artifact
```
Comment (sfuller) [inline src/MonC/Artifact.cs]: I don't see this being used anywhere?

File: src/MonC/Codegen/ILListingWriter.cs [hunk 3]
```diff
@@ -1,18 +1,17 @@
...
 using System.IO;
 using System.Linq;
 using MonC.Bytecode;
-using MonC.Codegen;
 
-namespace MonC.Frontend
+namespace MonC.Codegen
 {
-    public class IntermediateLanguageWriter
+    public class ILListingWriter
```
Comment (sfuller) [inline src/MonC/Codegen/ILListingWriter.cs:9]: This class name really demonstrates the shortcomings of the defacto C# interface naming scheme :P

File: src/MonC/Artifact.cs (hunk 2 above)
Comment (jackoalan) [reply src/MonC/Artifact.cs]: This whole file is actually backported from the llvm branch where the new Driver is being implemented.

It's used here:
https://github.com/jackoalan/MonC/blob/llvm/src/Driver/JobAction.cs#L42

I've yet to actually implement the Linker and VM tool in the Driver, but I can backport the artifact interface to Frontend once that's figured out

**Pull request merged**